[HarishB99] iP

HarishB99 opened the pull request:
# Bhaymax

> "I am a robot. I cannot be offended." - Baymax ([_Big Hero 6 (2014)_](https://www.google.com/search?q=big+hero+6&oq=big+hero&gs_lcrp=EgZjaHJvbWUqEggAEAAYQxjjAhixAxiABBiKBTISCAAQABhDGOMCGLEDGIAEGIoFMg8IARAuGEMYsQMYgAQYigUyBggCEEUYOTIMCAMQABhDGIAEGIoFMgwIBBAAGEMYgAQYigUyDAgFEAAYQxiABBiKBTIMCAYQLhhDGIAEGIoFMgwIBxAAGEMYgAQYigUyDAgIEC4YQxiABBiKBTIMCAkQLhhDGIAEGIoF0gEIMzYzM2owajeoAgCwAgA&sourceid=chrome&ie=UTF-8))

Bhaymax is designed to be a simple yet efficient task manager. It's

* text-based
* easy to learn
* ~FAST~ _SUPER_ FAST to use

All you need to do is,

1. Download it from [here](https://github.com/HarishB99/ip/releases/download/v0.1/bhaymax.jar)
2. Execute it from the command line
3. Add your tasks
4. Let it manage your tasks for you 🤖

And it is __FREE__!

Features:

- [x] Managing tasks
- [x] Managing deadlines
- [x] Filter by dates
- [x] Search by keyword
- [ ] Case-insensitive search (coming soon)
- [ ] Sort by date (coming soon)
- [ ] Group by date (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:

```java
public class Main {
    public static void main(String[] args) {
        new Bhaymax.run();
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

HarishB99 committed:
Rename chatbot to Bhaymax

HarishB99 committed:
Bhaymax.java: Print greeting, farewell messages and exit

HarishB99 committed:
EXPECTED.TXT: Change expected output to reflect the new program

HarishB99 committed:
Bhaymax.java: Add indentation to output

HarishB99 committed:
Bhaymax.java: Read in command from user and echo it back

HarishB99 committed:
Implement storing and listing of commands by user

Bhaymax.java: Add LinkedList to store commands from user
Bhaymax.java: Implement list command

HarishB99 committed:
feat: Add ability to mark tasks as done/undone

Task.java: Create Task class to store tasks
Bhaymax.java: Implement commands to mark tasks as done/undone

HarishB99 committed:
feat: Add three types of events

Todo.java: Create Todo class from Task class
Deadline.java: Create Deadline class from Task class
Event.java: Create Event class from Task class
Bhaymax.java: Add commands to create the three new types of tasks

HarishB99 committed:
Baymax.java: Create and use static method printWithIndent

HarishB99 committed:
Add a simple test case

HarishB99 committed:
Bhaymax.java: Replace sysout with printWithIndent

HarishB99 committed:
Add test case for adding and listing different kinds of tasks

HarishB99 committed:
Cleanup code for better readability

Code is messy due to if-else constructs

Use switch statements to improve readability of the code

Include additional exceptions for clearer error handling

Tokenise user input instead of splitting using separator

HarishB99 committed:
Handle errors

HarishB99 committed:
Add test cases for blank input and erraneous inputs

HarishB99 committed:
fix: Add test cases that were not added in previous commit

HarishB99 committed:
Implement delete tasks

HarishB99 committed:
fix: Add changes to Bhaymax.java which was not added in previous commit

HarishB99 committed:
Bhaymax.java: Indent code to follow code standards

HarishB99 committed:
Customise error messages

Error messages were too generic

Change error messages to be more specific
Change test cases accordingly
Fix: Change test scripts to use all test cases

There will be better clarity on what went
wrong when something fails

HarishB99 committed:
Implement read file

HarishB99 committed:
Implement write file

HarishB99 committed:
Bhaymax.java: Save tasks to file on change

HarishB99 committed:
Merge branch 'branch-Level-7'

HarishB99 committed:
Store dates as LocalDateTime instead of String

HarishB99 committed:
Implement task filtering by date and/or time

HarishB99 committed:
Merge branch 'branch-Level-8'

HarishB99 committed:
wip: Extract functionalities to their own classes

All functionalities were grouped together within
the main method in Bhaymax.java. This made the code
look very messy.

Follow OOP principles
Extract Commands to their own classes
Extract Parser to its own class
Extract Storage to its own class
Extract TaskList to its own class
Extract Ui to its own class

The code is more well-organised and easier to maintain

HarishB99 committed:
Event.java: fix typo end -> start

HarishB99 committed:
Parser.java: fix parsing of todo command

HarishB99 committed:
Ui.java: Modify error message for DateTimeParseException for accuracy

HarishB99 committed:
fix: Create data directory if missing

HarishB99 committed:
fix: Modify test cases to reflect changes in codebase

HarishB99 committed:
runtest.bat: (wip) Change runtest.bat to remove data directory before each run

HarishB99 committed:
Add test case to mark, unmark, delete tasks when list is empty

HarishB99 committed:
Merge branch 'branch-A-MoreOOP'

HarishB99 committed:
Bhaymax.java: Remove unnecessary constant TASKS_FILE_PATH

HarishB99 committed:
Move Exception classes to bhaymax.exception package

HarishB99 committed:
Move classes to packages

HarishB99 committed:
Merge branch 'branch-A-Packages'

HarishB99 committed:
Merge branch 'add-gradle-support'

HarishB99 committed:
build.gradle: Configure name and main class for gradle

HarishB99 committed:
Add checkstyle plugin to gradle

HarishB99 committed:
Add test for TaskList::addTask

HarishB99 committed:
Add test for Parser::parse on empty string

HarishB99 committed:
Move Bhaymax.java into bhaymax package

HarishB99 committed:
Bhaymax.java: Rename constant NAME -> APP_NAME

HarishB99 committed:
wip: Add javadoc comments

All classes were missing JavaDoc comments

Add javadoc comments to the following:
- Classes under bhaymax.exception
- Classes under bhaymax.parser
- Classes under bhaymax.storage
- Classes under bhaymax.ui
- Classes under bhaymax.util
- Bhaymax.java

The above classes now have their public
methods documented for easier reference.

JavaDoc comments needs to be
added to the following classes:
- Classes under bhaymax.command
- Classes under bhaymax.task
- Classes under bhaymax.task.timesensitive

HarishB99 committed:
wip: Add javadoc to bhaymax.command.FilterOpt

HarishB99 committed:
Bhaymax.java: Organise imports to follow code standard

HarishB99 committed:
Ui.java: Organise imports to follow code standard

HarishB99 committed:
Storage.java: Organise code to follow code standard

HarishB99 committed:
Parser.java: Organise code to follow code standard

HarishB99 committed:
Organise code in bhaymax.task

HarishB99 committed:
Organise code in bhaymax.command

HarishB99 committed:
Add javadoc comments

    Several classes were missing JavaDoc comments

    Add javadoc comments to the following:
    - Classes under bhaymax.command
    - Classes under bhaymax.task
    - Classes under bhaymax.task.timesensitive

    Now all classes have their public
    methods documented for easier reference.

HarishB99 committed:
Merge branch 'branch-A-JavaDoc'

HarishB99 committed:
Merge branch 'branch-A-CodingStandard'

HarishB99 committed:
Add javadoc comments for stubs created

HarishB99 committed:
Change test codes to follow code standard

HarishB99 committed:
Implement search command (to search tasks by keyword)

HarishB99 committed:
Merge branch 'branch-Level-9'

HarishB99 committed:
fix: print the correct task index when filtering

HarishB99 committed:
Add javadoc comment for TaskList::printTasksContainingSearchTerm

HarishB99 committed:
TaskList.java: Remove unused import statement

HarishB99 committed:
Parser.java: Add 'exit' as another exit command

HarishB99 committed:
README.md: Change references to Duke to Bhaymax

Clifong reviewed (commented):
Personally I think the constants should be set as private. Alternatively, it is better to places these constants in another class that is purely just data, no methods.

Clifong reviewed (commented):
Good practice to remove unused import

Clifong reviewed (commented):
Nice use of streams

Clifong reviewed (commented):
Short and simple comments are good

HarishB99 committed:
Move command string constants to its own Enum

HarishB99 committed:
Merge branch 'branch-A-CommandEnums'

HarishB99 committed:
runtest.bat: fix: Include package name when running with java

HarishB99 committed:
runtest.sh: Include package name when running with java

HarishB99 committed:
runtest.sh: Build Bhaymax.java instead of *.java

HarishB99 committed:
build.gradle: Add dependencies for JavaFX

HarishB99 committed:
Move Bhaymax from bhaymax to bhaymax.main

HarishB99 committed:
build.gradle: Fix mainClass path for Bhaymax

carineang commented on src/main/java/bhaymax/ui/Ui.java:
```diff
@@ -0,0 +1,158 @@
+     *
+     * @param exception An {@code InvalidCommandFormatException} object
+     */
+    public void showError(InvalidCommandFormatException exception) {
```
Good approach for catching and handling different types of errors.

carineang commented on src/main/java/bhaymax/command/CommandString.java:
```diff
@@ -0,0 +1,61 @@
+ * the valid commands a user is allowed to
+ * enter
+ */
+public enum CommandString {
```
Good use of enums

carineang commented on src/main/java/bhaymax/task/timesensitive/TimeSensitiveTask.java:
```diff
@@ -0,0 +1,26 @@
+/**
+ * An interface for task types that need to track time
+ */
+public interface TimeSensitiveTask {
```
good use of interface

carineang reviewed (commented):
Generally well done :)

HarishB99 committed:
Implement GUI for Bhaymax chatbot

HarishB99 committed:
Combine responses into one single DialogBox

HarishB99 committed:
Merge branch 'branch-Level-10'

HarishB99 committed:
Add space after numbered bullet points when listing tasks

HarishB99 committed:
Main.java: Collapse catch blocks

HarishB99 committed:
Main.java: Make TaskList final

HarishB99 committed:
Main.java: Rectify comment to reflect behaviour of MainWindow

HarishB99 commented:
> Personally I think the constants should be set as private. Alternatively, it is better to places these constants in another class that is purely just data, no methods.

I agree that the constants are best placed in another class. I have now moved them into an Enumeration class. Thank you for the suggestion!

HarishB99 commented:
> Good practice to remove unused import

Hi! Just to clarify, did you notice any unused imports in my code?

HarishB99 commented:
> Short and simple comments are good

Hi @Clifong ! Just to clarify, do you think I should shorten my comments or do you think my comments are short and simple?

HarishB99 committed:
Move Main.java into ui package

HarishB99 committed:
Remove text-based UI

HarishB99 committed:
DialogBox.java: Throw RuntimeException when IOException occurs

HarishB99 committed:
Anchor TextField, Button and ScrollPane in MainWindow

HarishB99 committed:
Bhaymax.java: Add minimum and maximum dimensions for JavaFX window

HarishB99 committed:
Change and add new images for chatbot

HarishB99 committed:
MainWindow.java: Change Duke to Bhaymax in comments

HarishB99 committed:
MainWindow.java: Change variable name dialogBoxes -> responses

HarishB99 committed:
MainWindow.java: Remove unused function displayDialogs

HarishB99 committed:
Add and rename images

HarishB99 committed:
Personalise Bhaymax and improve code quality

Initially, messages from Bhaymax echoed the personality of Duke
instead of Baymax. Additionally, there were instances of unnecessarily long code.

The following changes were made to give Bhaymax his unique quality:
- Personalise chatbot responses from Bhaymax
- Change chatbot avatar images
- Additionally, change user avatar image for better quality

The following changes were made to improve the quality of the code
base:
- Lint comments
- Check if task matches date filter inside TimeSensitiveTask
  instead of within TaskList

Now, Bhaymax has more personality with his own responses and
expressions. The code quality is also better, though it can still be
improved.

HarishB99 committed:
feat: Add clear command

HarishB99 committed:
Add 'hello' and 'hi' commands

HarishB99 committed:
Parser.java: Accept case-insensitive commands

HarishB99 committed:
TaskListTest.java: Add test cases for addTask and isValidIndex

HarishB99 committed:
ParserTest.java: Add test cases for trivial commands

Parsing trivial commands were not well tested

Add test cases for parsing the following trivial commands
- bye
- exit
- hi
- hello
- clear

Now test cases have been added to test
the parsing of the above commands

HarishB99 committed:
build.gradle: Enable assertions

HarishB99 committed:
Display error when file format is corrupted/invalid

HarishB99 committed:
Move Bhaymax.java from bhaymax.ui bhaymax.main

HarishB99 committed:
Bhaymax.java: Rectify order of showing welcome dialog and setting tasks

HarishB99 committed:
Parser.java: Raise exception when fields in the middle are missing

HarishB99 committed:
Merge branch 'master' into branch-A-MoreTesting

HarishB99 committed:
wip: Organise and cleanup constants into enumerations

HarishB99 committed:
Parser.java: Refactor code by extracting out functions

HarishB99 committed:
MainWindow.java: Show responses for error from app with sad face

HarishB99 committed:
wip: Improve code quality

HarishB99 committed:
fix: Change regexes to read tasks from file

HarishB99 committed:
wip: Make InvalidCommandException a child of InvalidCommandFormatException

HarishB99 committed:
Move magic values into constants and perform extract refactoring

HarishB99 committed:
SearchCommand.java: Remove ununsed attributes

HarishB99 committed:
wip: Improve code quality

Code quality still had room for improvement

- Move more magic values into constants
- Rectify/beautify comments
- Remove commented out codes
- Remove EMPTY_STRING constant

Code quality is better, but it can still be improved
in the following ways
- Move magic values in *Command classes into constants
- Organise constants to remove duplicates and have more clarity
- Scan through code base for signs of weak SLAP, long methods
  and nested codes

HarishB99 committed:
Move string literals in Command classes into constants

HarishB99 committed:
Move constants and create exceptions for better clarity

HarishB99 committed:
Move error message constants into exceptions

HarishB99 committed:
Split InvalidDateFormatException into two

InvalidDateFormatException was responsible for two exceptional cases
1. Invalid date formats in commands
2. Invalid date formats in files
This made it hard to track within the program where the exception
occurred.

Split the exception into one for in command and another for in file

Now it is possible to 'track' where the exception occurred and
respond to the user accordingly

HarishB99 committed:
Rename DATE_FORMAT -> DATE_INPUT_FORMAT

HarishB99 committed:
Move error message literals into exceptions for tasks

HarishB99 committed:
Move error message literal from Storage into an exception

HarishB99 committed:
DeadlineCommand.java: Change comment to be more accurate

HarishB99 committed:
Move literal flag constants to their own command classes

HarishB99 committed:
Task.java: Remove unnecessary methods

HarishB99 committed:
Launcher.java: Change comment for brevity and clarity

HarishB99 committed:
Cleanup code and view for MainWindow

HarishB99 committed:
Move string literals to constants in Exception classes

HarishB99 committed:
Improve code quality for Task related classes

HarishB99 committed:
Storage.java: Improve code quality (comments + code)

HarishB99 committed:
Parser.java: Improve code quality (add constants)

HarishB99 committed:
Change function names in MainWindow.java for better clarity

HarishB99 committed:
Improve code quality for comments in Command classes

HarishB99 committed:
Merge pull request #2 from HarishB99/branch-A-CodeQuality

Improve code quality

HarishB99 committed:
DeleteCommand.java: Assert number of remaining tasks is non-negative

HarishB99 committed:
Add assertion and extract out methods in Task classes

HarishB99 committed:
Merge pull request #3 from HarishB99/branch-A-Assertions

Add assertions

HarishB99 committed:
Add support for Github Actions

HarishB99 committed:
Bhaymax.java: Add app icon and window title

HarishB99 committed:
Add CSS

HarishB99 committed:
Implement flipped text bubbles

HarishB99 committed:
Personalise colors

HarishB99 committed:
Add background image for chatbox

HarishB99 committed:
Ignore empty user input

HarishB99 committed:
Remove maximum dimension restraint on window

HarishB99 committed:
Perform case insensitive search

HarishB99 committed:
Merge branch 'master' into branch-A-BetterGui

HarishB99 committed:
Implement sorting of tasks

Tasks were not sorted when displayed

Implement sorting of tasks
Throw error when start date is after end date for events

Tasks are now sorted by time first, followed by their description.
Additionally, dates ranges of events are validated so that
events can be sorted simply by their start time

HarishB99 committed:
fix: Event.java: Print date and time in output format

HarishB99 committed:
Merge branch 'master' into branch-A-CI

HarishB99 committed:
Merge pull request #4 from HarishB99/branch-A-CI

Configure CI via Github Actions

HarishB99 committed:
Merge branch 'master' into branch-A-BetterGui

HarishB99 committed:
Merge branch 'master' into branch-C-Sort

HarishB99 committed:
TaskList.java: Remove unused import Comparator

HarishB99 committed:
Detect and handle attempts to add duplicate tasks

HarishB99 committed:
TodoCommand.java: Remove unused import

HarishB99 committed:
Merge pull request #5 from HarishB99/branch-C-DetectDuplicates

Detect and handle attempts to add duplicate tasks

HarishB99 committed:
Merge branch 'master' into branch-A-BetterGui

HarishB99 committed:
Merge branch 'branch-A-BetterGui' into branch-A-MoreTesting

HarishB99 committed:
Use mockito library in place of stubs

HarishB99 committed:
Use ParameterizedTest for test cases with multiple test inputs

HarishB99 committed:
Change date input format 'dd-MM-yyyy' -> 'dd/MM/yyyy'

HarishB99 committed:
ParserTest.java: Rename test cases for better clarity

HarishB99 committed:
ParserTest.java: Add test cases for list command

HarishB99 committed:
ParserTest.java: Add test cases for parsing search command

HarishB99 committed:
Include command format in error message when missing search term

HarishB99 committed:
MainWindow.java: Show actual input from user instead of trimmed one

HarishB99 committed:
Merge branch 'branch-A-MoreErrorHandling'

HarishB99 committed:
Merge branch 'master' into branch-A-MoreTesting

HarishB99 committed:
Add test case for parsing todo command

HarishB99 committed:
Add test cases for parsing mark, unmark and delete commands

HarishB99 committed:
ParserTest.java: Add test case when task number provided is a fraction

HarishB99 committed:
Parser.java: Pass exception to throw as argument

Exceptions to throw were decided based on conditionals,
which introduced SLAP problems.

Pass exception to throw in method as argument

Now, the exception to throw is determined within the
switch case instead. This fixes the SLAP problem and
improves readability and maintainability.

HarishB99 committed:
Parser.java: Remove unused parameter commandString in getTaskIndex

HarishB99 committed:
Parser.java: Rename variables extracted from deadline command

HarishB99 committed:
Organise exceptions into sub packages

HarishB99 committed:
wip: Add test cases for commands with dates

HarishB99 committed:
wip: Update documentation of program

HarishB99 committed:
README.md: Add descriptions for trivial commands

HarishB99 committed:
docs/README.md: Reformat file

HarishB99 committed:
Merge branch 'branch-A-MoreTesting' into branch-A-BetterGui

HarishB99 committed:
Dynamically style dialog boxes

HarishB99 committed:
Merge branch 'branch-A-BetterGui' into branch-A-MoreTesting

HarishB99 committed:
DialogBoxStyleClass.java: Remove unused imports

HarishB99 committed:
docs/README.md: Document commands and their formats

HarishB99 committed:
Add docs/Ui.png

HarishB99 committed:
Merge pull request #6 from HarishB99/branch-A-MoreTesting

Add documentation

HarishB99 committed:
Change Ui.png to include product name

HarishB99 committed:
Improve documentation

Documentation was missing usage instructions and
examples for filter command.

docs/README.md: Add examples for filter command
docs/README.md: Add usage instructions for chatbot

The added instructions and examples should help
to improve the clarity and completeness of the
documentation.

HarishB99 committed:
Merge pull request #7 from HarishB99/branch-A-UserGuide

Improve documentation

HarishB99 committed:
README.md: Change Main class path Bhaymax -> bhaymax.main.Launcher

HarishB99 committed:
README.md: Include link to original ip repo
